Add support to schedule operations

brunolm opened the pull request:
Closes #69 

# Changes

- Added:
    - `shared/Operations`: class that contains all functions that can be scheduled. The function must receive the parameters: `lotus` (an instance of the lotus class), `metadata` (information saved when the operation was queued; e.g. deal)
    - `shared/OperationsQueue`: class that handles the queue and execution of the operations
    - `popup/

# How to use

### Adding operations to the queue:

1. Add a function you'd like to execute on `shared/Operations.js` file
1. Add the operation to the queue using:
```ts
// package to handle dates
import { DateTime } from 'luxon';

// queue instance
import { operationsQueue } from 'src/shared/OperationsQueue';

operationsQueue.queue({
  label: 'the name of the operation here, anything',

  // name of the method implemented in shared/Operations
  // hint: VSCode should show you the available names through its built-in TypeScript support
  f: 'youFnNameAsInOperationsFile',

  // object with anything you want to store to pass to the operation later (e.g. `deal`)
  metadata: {},

  // time when to invoke the function
  invokeAt: DateTime.local().plus({ hours: 12 }).toString(),
});
```

# Preview
![operations](https://user-images.githubusercontent.com/706078/96900506-8e362700-1468-11eb-8ba1-e83640fcd2de.gif)

brunolm committed:
Add support to schedule operations

brunolm committed:
Add better status; sort by exec time

brunolm committed:
Remove test calls

brunolm committed:
Add retry support; Add remove method

mgoelzer commented:
@brunolm 

This worked _almost_ perfectly.  However, I found one bug that I hope you can fix before we merge.

## Error:  `Invalid asm.js: Unexpected Token`

The setup here is that I checkout your `issue-69/scheduled-operations` branch and do `yarn && yarn build` without any code modifications.  Then I get the error.

[This video shows the error](https://drive.google.com/file/d/1Uz6Hx7ui3E3euax_wVVJ9goykWpExfkV/view?usp=sharing
).  

I don't want to merge with this error unfixed because more code will get piled on top and then it'll be hard to find the root cause later.

## Same error in firefox?

I think the same error is showing up in the Firefox logs:

<img width="1145" alt="Screen Shot 2020-10-22 at 9 29 08 PM" src="https://user-images.githubusercontent.com/225111/96953144-a0868400-14ad-11eb-83f0-433ad15127ee.png">

To see the startup logs, use `Inspect` button instead of Reload:

<img width="721" alt="Debugging_-_Runtime___this-firefox" src="https://user-images.githubusercontent.com/225111/96953236-d75c9a00-14ad-11eb-8bac-55a953dd4633.png">

brunolm committed:
Fix lint errors

brunolm commented:
@mgoelzer 

I pushed some changes to fix lint issues and I also added a new script on package.json to run lint (`yarn lint`)

About the other issues:

1. `Invalid asm.js: Unexpected Token`
    1. I cloned the project in a new folder
    1. ran `yarn && yarn build`
    1. switched to this branch
    1. `yarn && yarn build`
    1. did not see any errors

2. This video shows the error.
    1. the video is outdated, it doesn't have the checkbox and the miner field is not disabled, if the latest version if you want to get a file from the browser you can't specify a miner id

brunolm commented on src/popup/App/Home/OperationsList/OperationsList.js:
```diff
@@ -0,0 +1,96 @@
+/* global chrome */
```
Because we don't have TypeScript here we are required to tell the code that `chrome` variable does exist globally. However this file is not using the global variable so it makes the linter fail. I've removed this line.

I also added a new script on package.json, you can check for lint errors with `yarn lint`

brunolm committed:
Fix format

brunolm commented:
Mike:
> Invalid asm.js: Unexpected Token
> Happens on the first fresh install

mgoelzer commented:
@brunolm 

> Happens on the first fresh install

Yes, it happens on fresh install, but then it keeps happening forever until I fully uninstall it and install master or one of the other branches.

brunolm committed:
Workaround Invalid asm.js: Unexpected token error

mgoelzer commented:
LGTM
